File: 383.RansomNote/Note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> letterCount = new HashMap<>();
        for (char c : magazine.toCharArray()) {
            letterCount.put(c, letterCount.getOrDefault(c, 0) + 1);
        }

        for (char c : ransomNote.toCharArray()) {
            if (letterCount.getOrDefault(c, 0) <= 0) {
                return false;
            }
            letterCount.put(c, letterCount.getOrDefault(c, 0) - 1);
        }

        return true;
    }
}

File: 383.RansomNote/note.cc
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        map<char, int> letterCount;
        for (char c : magazine) {
            letterCount[c]++;
        }

        for (char c : ransomNote) {
            if (letterCount[c] <= 0) {
                return false;
            }
            letterCount[c]--;
        }

        return true;
    }
};
